protocols/bftcosi/bftcosi.go

```diff
@@ -97,23 +102,46 @@ type ProtocolBFTCoSi struct {
 
 	// finale signature that this BFTCoSi round has produced
 	finalSignature *BFTSignature
+
+	// our index in the Roster list
+	index int
+	// prepareSignature is the signature generated during the prepare phase
+	// This signature is adapted according to the exceptions that occured during
+	// the prepare phase.
+	prepareSignature []byte
+	// commitSignature is the signature generated during the commit phase
+	// This signature is adapted according to the exceptions that occured during
+	// the commit phase.
+	commitSignature []byte
+
+	// commitCommitDone is to ensure the Challenge phase of the Commitment phase
+	// (the second commit) is done, before doing the challenge.
+	// It's needed because both phases are started *almost* simultaneously and
+	// the commit phase can be de-sync.
+	commitCommitDone chan bool
 }
 
 // NewBFTCoSiProtocol returns a new bftcosi struct
 func NewBFTCoSiProtocol(n *sda.TreeNodeInstance, verify VerificationFunction) (*ProtocolBFTCoSi, error) {
 	// initialize the bftcosi node/protocol-instance
 	bft := &ProtocolBFTCoSi{
-		TreeNodeInstance: n,
-		suite:            n.Suite(),
-		prepare:          cosi.NewCosi(n.Suite(), n.Private()),
-		commit:           cosi.NewCosi(n.Suite(), n.Private()),
-		verifyChan:       make(chan bool),
-		doneProcessing:   make(chan bool, 2),
-		doneSigning:      make(chan bool, 1),
-		verificationFun:  verify,
-		AggregatedPublic: n.Roster().Aggregate,
-		threshold:        int(2.0 * math.Ceil(float64(len(n.Tree().List()))/3.0)),
-	}
+		TreeNodeInstance:     n,
+		suite:                n.Suite(),
+		prepare:              cosi.NewCosi(n.Suite(), n.Private(), n.Roster().Publics()),
+		commit:               cosi.NewCosi(n.Suite(), n.Private(), n.Roster().Publics()),
+		verifyChan:           make(chan bool),
+		doneProcessing:       make(chan bool, 2),
+		doneSigning:          make(chan bool, 1),
+		VerificationFunction: verify,
+		AggregatedPublic:     n.Roster().Aggregate,
+		threshold:            len(n.Tree().List()) * 2 / 3,
```

Comment:
I wrote a test for it `TestThreshold` - please have a look if it's OK. I'm a bit ashamed it took me more than 5 minutes to do the integer-math. +45 minutes for getting the test to run ;)